doc/reference/trd-syscalls.md

```diff
@@ -0,0 +1,715 @@
+buffer must be readable and writeable by the process. Zero-length
+buffers may therefore have abitrary addresses. If the passed buffer is
+not complete within the calling process's readable address space, the
+kernel MUST return a failure result with an error code of INVAL
+(invalid value).
+
+4.6 Memop (Class ID: 5)
+---------------------------------
+
+The Memop class is how a userspace process requests and provides
+information about its address space.  The register arguments for
+Allow system calls are as follows. The registers r0-r3 correspond
+to r0-r3 on CortexM and a0-a3 on RISC-V.
+
+| Argument               | Register |
+|------------------------|----------|
+| Operation              | r0       |
+| Operation argument     | r1       |
+| unused                 | r2       |
+| unused                 | r3       |
+
+The operation argument specifies which memory operation to perform. There
+are 12:
+
+| Memop Operation | Operation                                               | Success          |
+|-----------------|---------------------------------------------------------|------------------|
+| 0               | Break                                                   | Success          |
+| 1               | SBreak                                                  | Success with u32 |
+| 2               | Get process RAM start address                           | Success with u32 |
+| 3               | Get address immediately after process RAM allocation    | Success with u32 |
+| 4               | Get process flash start address                         | Success with u32 |
+| 5               | Get address immediately after process flash region      | Success with u32 |
+| 6               | Get lowest address (end) of the grant region            | Success with u32 |
+| 7               | Get number of writeable flash regions in process header | Success with u32 |
+| 8               | Get start address of a writeable flash region           | Success with u32 |
+| 9               | Get end adddress of a writeable flash region            | Success with u32 |
+| 10              | Set the start of the process stack                      | Success          |
+| 11              | Set the start of the process heap                       | Success          |
+
+The success return variant is Memop class system call specific and
+specified in the table above. All Memop class system calls have a
+_Failure_ failure type.
+
+4.7 Exit (Class ID: 6)
+--------------------------------
+
+The Exit system call class is how a userspace process terminates. 
+Successful calls to Exit system calls do not return.
+
+There are two Exit system calls:
+  - `exit-terminate`
+  - `exit-restart`
+
+The first call, `exit-terminate`, terminates the process and tells the
+kernel that it may reclaim and reallocate the process as well as all of its
+resources. Usually this indicates that the process has completed
+its work. 
```

Comment:
Perhaps we should say something about how resources from a terminated process MAY be reused by future processes? We're explicit about this in the restart case, but it's perhaps ambiguous here whether process resources are cleared if a process is not restarted (they are currently not, but we also don't have too dynamic a process setup just yet).

My take here is:
 - There will be value to some users in the future in making guarantees about what happens to process resources once a process is permanently terminated
 - It's not totally clear what promises we will want to make. Rather than trying to sort that out right now, we should be explicit that this "basic" terminate syscall makes no promises. We can add a future "secure terminate" (or possibly have a Secure Process type that implicitly does a secure terminate, or probably other ideas) when it's clearer what the exact needs and use will be.

Reply:
#2431

Reply:
> Perhaps we should say something about how resources from a terminated process MAY be reused by future processes? 

Like what? The process is dead, it shouldn't care. The MAY statement in restart was so that applications don't assume they are in the same location in memory when they restart. E.g., an application should not write memory addresses to non-volatile memory, reboot, and assume those addresses are still valid. Making it explicit that an application can't assume this allows the kernel to do a compacting garbage collection on process memory.

> We're explicit about this in the restart case, but it's perhaps ambiguous here whether process resources are cleared if a process is not restarted

I think it's ambiguous because it's up to the kernel.

>     * There will be value to some users in the future in making guarantees about what happens to process resources once a process is permanently terminated

What guarantees should they be? I am wary about making blanket guarantees because of the diversity of use cases. E.g., we don't want to require that a kernel has code for process garbage collection if it doesn't expect to have dynamic processes.

>     * It's not totally clear what promises we will want to make. Rather than trying to sort that out right now, we should be explicit that this "basic" terminate syscall makes no promises. We can add a future "secure terminate" (or possibly have a Secure Process type that implicitly does a secure terminate, or probably other ideas) when it's clearer what the exact needs and use will be.

This kind of seems backwards. By stating that there are no promises., we can't later write that there are promises if we figure out it's something important. I.e., we might need to deprecate Exit. However, if we leave it unsaid, this (from a specification standpoint) means that the kernel is unconstrained. We can then leave it open, or later decide to specify it if need be.

How would "Kernel policies and mechanisms for reclaiming or reusing process resources is outside the scope of this document." sound?

Reply:
I'm tracking changes in #2431